pkg/network/usm/monitor_test.go

```diff
@@ -808,3 +814,360 @@ func TestCleanProtocolMaps(t *testing.T) {
+	return func(t *testing.T, conn net.Conn) {
+		_, err := conn.Write(buffer)
+		require.NoError(t, err)
+		_, err = conn.Read(buffer)
+		require.NoError(t, err)
+	}
+}
+
+type connectionStatesMapTestCase struct {
+	// name is the name of the test
+	name string
+	// cfg is the USM configuration to use for the test
+	cfg *networkConfig.Config
+	// preTestSetup is a function that will be called before sending requests to the server, but
+	// after monitor and client initialization. It can allow us to manipulate other USM maps, such as
+	// connection_protocol to influence the test
+	preTestSetup func(*testing.T, *Monitor, net.Conn)
+	// useTLS indicates whether to use TLS for the connection
+	useTLS bool
+	// expectedResult indicates whether the connection should exist in the connection_states map
+	expectedResult existenceResult
+	// sendRequestCallback is a function that will be called to send the request to the server.
+	// the method is responsible for reading the response from the server.
+	sendRequestCallback func(*testing.T, net.Conn)
+	// skipCondition is a function that will be called to determine whether to skip the test
+	skipCondition func(*testing.T)
+}
+
+func testConnectionStatesMap(t *testing.T, testParams connectionStatesMapTestCase) {
+	if testParams.skipCondition != nil {
+		testParams.skipCondition(t)
+	}
+	monitor := setupUSMTLSMonitor(t, testParams.cfg, useExistingConsumer)
+
+	serverConns := make([]net.Conn, 0)
+	t.Cleanup(func() {
+		closeConnections(serverConns...)
+	})
+	// echo server
+	srvFn := func(conn net.Conn) {
+		// We intentionally don't close the connection, to ensure the connection won't be deleted
+		// from the `connection_states` map
+		serverConns = append(serverConns, conn)
+		// Echo back whatever is received
+		_, _ = io.Copy(conn, conn)
+	}
+	srv := testutil.NewTCPServer("127.0.0.1:0", srvFn, testParams.useTLS)
+	done := make(chan struct{})
+	require.NoError(t, srv.Run(done))
+	t.Cleanup(func() { close(done) })
+
+	var clientConn net.Conn
+	var err error
+	if testParams.useTLS {
+		clientConn, err = tls.Dial("tcp", srv.Address(), &tls.Config{InsecureSkipVerify: true})
+	} else {
+		clientConn, err = net.DialTimeout("tcp", srv.Address(), 5*time.Second)
+	}
+	require.NoError(t, err)
+	t.Cleanup(func() { closeConnections(clientConn) })
+
+	if testParams.preTestSetup != nil {
+		testParams.preTestSetup(t, monitor, clientConn)
+	}
+
+	if testParams.sendRequestCallback != nil {
+		testParams.sendRequestCallback(t, clientConn)
+	} else {
+		sendAndReadBuffer(randomBuffer)(t, clientConn)
+	}
+
+	m, _, err := monitor.ebpfProgram.GetMap("connection_states")
+	require.NoError(t, err)
+
+	checkConnExistenceInConnectionStatesMap(t, m, clientConn, testParams.expectedResult)
+
+	// Close the connection and expect it to be removed from the map
+	closeConnections(serverConns...)
+	closeConnections(clientConn)
+	// Wait for the connection to be removed from the map
+	time.Sleep(100 * time.Millisecond)
```

Comment:
maybe using Eventually instead?

Reply:
since we're checking something is not presenting in a map, we cannot use `require.Eventually` as it might be flaky